File: android/src/main/java/net/xpece/android/view/XpView.java
package net.xpece.android.view;

import android.animation.LayoutTransition;
import android.annotation.TargetApi;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import net.xpece.android.AndroidUtils;
import net.xpece.android.R;

/**
 * Created by pechanecjr on 4. 1. 2015.
 */
@TargetApi(21)
public final class XpView {
    protected XpView() {}

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    @SuppressWarnings("deprecation")
    public static void setBackground(View v, Drawable d) {
        if (Build.VERSION.SDK_INT < 16) {
            v.setBackgroundDrawable(d);
        } else {
            v.setBackground(d);
        }
    }

    public static void setVisibility(View v, boolean visible) {
        v.setVisibility(visible ? View.VISIBLE : View.GONE);
    }

    public static void setTextAndVisibility(TextView v, CharSequence text) {
        boolean visible = !TextUtils.isEmpty(text);
        if (visible) {
            v.setText(text);
            v.setVisibility(View.VISIBLE);
        } else {
            v.setVisibility(View.GONE);
            v.setText(null);
        }
    }

    /**
     * @return Returns true this ScrollView can be scrolled
     */
    public static boolean canScroll(ScrollView scroll) {
        View child = scroll.getChildAt(0);
        if (child != null) {
            int childHeight = child.getHeight();
            return scroll.getHeight() < childHeight + scroll.getPaddingTop() + scroll.getPaddingBottom();
        }
        return false;
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    @SuppressWarnings("deprecation")
    public static void removeOnGlobalLayoutListener(View v, ViewTreeObserver.OnGlobalLayoutListener l) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
            v.getViewTreeObserver().removeGlobalOnLayoutListener(l);
        } else {
            v.getViewTreeObserver().removeOnGlobalLayoutListener(l);
        }
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public static void setSearchViewLayoutTransition(android.widget.SearchView view) {
        if (!AndroidUtils.API_11) return;
        int searchBarId = view.getContext().getResources().getIdentifier("android:id/search_bar", null, null);
        LinearLayout searchBar = (LinearLayout) view.findViewById(searchBarId);
        searchBar.setLayoutTransition(new LayoutTransition());
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public static void setSearchViewLayoutTransition(android.support.v7.widget.SearchView view) {
        if (!AndroidUtils.API_11) return;
        LinearLayout searchBar = (LinearLayout) view.findViewById(R.id.search_bar);
        searchBar.setLayoutTransition(new LayoutTransition());
    }

}

File: android/src/main/java/net/xpece/android/view/XpView.kt
@file:JvmName("XpView")

package net.xpece.android.view

import android.animation.LayoutTransition
import android.annotation.TargetApi
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.TransitionDrawable
import android.os.Build
import android.support.annotation.DrawableRes
import android.support.v4.view.ViewCompat
import android.support.v7.widget.AppCompatDrawableManager
import android.view.View
import android.view.ViewTreeObserver
import android.widget.*
import net.xpece.android.R

/**
 * @author Eugen on 29.07.2016.
 */

fun View.gone(): View {
    visibility = View.GONE
    return this
}

fun View.invisible(): View {
    visibility = View.INVISIBLE
    return this
}

fun View.visible(): View {
    visibility = View.VISIBLE
    return this
}

fun View.isVisible(): Boolean = visibility == View.VISIBLE

fun View.setVisible(visible: Boolean) = if (visible) visibility = View.VISIBLE else visibility = View.GONE

fun TextView.setTextAndVisibility(text: CharSequence?, invisible: Boolean = false) {
    if (text.isNullOrBlank()) {
        if (invisible) invisible() else gone()
        setText(null)
    } else {
        setText(text)
        visible()
    }
}

@TargetApi(16)
@Suppress("deprecation")
fun View.setBackgroundCompat(d: Drawable) {
    if (Build.VERSION.SDK_INT < 16) {
        setBackgroundDrawable(d)
    } else {
        background = d
    }
}

@TargetApi(Build.VERSION_CODES.JELLY_BEAN)
@Suppress("deprecation")
fun View.removeOnGlobalLayoutListenerCompat(l: ViewTreeObserver.OnGlobalLayoutListener) {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
        viewTreeObserver.removeGlobalOnLayoutListener(l)
    } else {
        viewTreeObserver.removeOnGlobalLayoutListener(l)
    }
}

/**
 * @return Returns true this ScrollView can be scrolled
 */
fun ScrollView.canScroll(): Boolean {
    val child = getChildAt(0)
    if (child != null) {
        val childHeight = child.height
        return height < childHeight + paddingTop + paddingBottom
    }
    return false
}

@TargetApi(Build.VERSION_CODES.HONEYCOMB)
fun setSearchViewLayoutTransition(view: SearchView) {
    if (Build.VERSION.SDK_INT < 11) return
    val searchBarId = view.context.resources.getIdentifier("android:id/search_bar", null, null)
    val searchBar = view.findViewById(searchBarId) as LinearLayout
    searchBar.layoutTransition = LayoutTransition()
}

@TargetApi(Build.VERSION_CODES.HONEYCOMB)
fun setSearchViewLayoutTransition(view: android.support.v7.widget.SearchView) {
    if (Build.VERSION.SDK_INT < 11) return
    val searchBar = view.findViewById(R.id.search_bar) as LinearLayout
    searchBar.layoutTransition = LayoutTransition()
}

fun ImageView.switchImage(d: Drawable, duration: Int = 100) {
    if (ViewCompat.isLaidOut(this)) {
        var old = drawable
        if (old is TransitionDrawable) {
            old = old.getDrawable(1)
        } else if (old == null) {
            old = ColorDrawable(0)
        }
        val d2 = TransitionDrawable(arrayOf(old, d))
        d2.isCrossFadeEnabled = true
        setImageDrawable(d2)
        d2.startTransition(duration)
    } else {
        setImageDrawable(d)
    }
}

fun ImageView.switchImage(@DrawableRes resId: Int, duration: Int = 100) {
    val d = AppCompatDrawableManager.get().getDrawable(context, resId)
    switchImage(d, duration)
}
